ceph/mds.go

```diff
@@ -0,0 +1,173 @@
+package ceph
+
+import (
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	cephCmd                      = "/usr/bin/ceph"
+	mdsBackgroundCollectInterval = 5 * time.Minute
+)
+
+const (
+	MDSModeDisabled   = 0
+	MDSModeForeground = 1
+	MDSModeBackground = 2
+)
+
+type mdsStat struct {
+	FSMap struct {
+		Filesystems []struct {
+			MDSMap struct {
+				FSName string `json:"fs_name"`
+				Info   map[string]struct {
+					GID   uint   `json:"gid"`
+					Name  string `json:"name"`
+					Rank  int    `json:"rank"`
+					State string `json:"state"`
+				} `json:"info"`
+			} `json:"mdsmap"`
+		} `json:"filesystems"`
+	} `json:"fsmap"`
+}
+
+// runMDSStat will run mds stat and get all info from the MDSs within the ceph cluster.
+func runMDSStat(config, user string) ([]byte, error) {
+	var (
+		out []byte
+		err error
+	)
+
+	if out, err = exec.Command(cephCmd, "-c", config, "-n", fmt.Sprintf("client.%s", user), "mds", "stat", "--format", "json").Output(); err != nil {
```

Comment:
timeouts on the command? esp if the MDS has a fault or there is network issues.

Reply:
/usr/bin/ceph has no timeouts on command sadly.

Reply:
If the container has the `timeout` command; you can implement as `timeout N ceph -c $config ...`

Consider how it can get stuck, and see if you need to tell timeout to explicitly SIGINT/TERM/KILL

Reply:
Ah wait, you mean for exec, yes, good point! Let me add them.

Reply:
Addressed.